### PR TITLE
Fix vmap constant output size

### LINKED ISSUE #1516: [BUG] vmap over function returning constant value has incorrect output shape
**Describe the bug**
The output shape of a vmapped function that returns a constant value is incorrect. A vmapped function that returns a constant value (i.e., the value does not depend on its arguments) should return an array with the same size at `out_axes` as the arguments specified in `in_axes`. I've included an example below.

**To Reproduce**
```python
import mlx.core as mx

@mx.vmap
def f1(x):
    return mx.array(1.0)

@mx.vmap
def f2(x):
    return x*0 + mx.array(1.0)

print(f1(mx.array([1, 2, 3])))  # array(1, dtype=float32)
print(f2(mx.array([1, 2, 3])))  # array([1, 1, 1], dtype=float32)
```

**Expected behavior**
The size of the output axis from a vmapped function should always equal the size of the arguments specified in `in_axes`. If the vectorized-over arguments have size 3 then the output shape of the function should also be 3 (at `out_axes`).

This ensures that functions returning constant values behave consistently, even when they do not depend on their inputs.

**Desktop (please complete the following information):**
 - OS Version: [e.g. MacOS 15.0.1]
 - Version [e.g. 0.19.0]

**Additional context**
Functions that return constant values are useful for maintaining a consistent interface, even if the output doesn't depend on the input. For example, in beamforming, different windowing functions (e.g., Hamming or rectangular) are applied to sensor data. While the rectangular window returns the same constant value (1) across all positions, it should still produce an output with the same shape as the input when used with vmap, ensuring compatibility with other windowing functions.

### PR DESCRIPTION
**Pull request by barronalex**
Resolves #1516

This mirror's Jax's `vmap` behaviour so I think it's reasonable.

### CODE REVIEW

File: mlx/transforms.cpp [hunk 1]
```diff
@@ -652,6 +652,17 @@ std::vector<array> vmap_replace(
     throw std::invalid_argument(msg.str());
   }
```
[the remark this answers is not in the document]
Comment (awni) [reply mlx/transforms.cpp:654]: Sorry I assumed tests failing meant this wasn't ready. I think if we go with this change then we should not allow all input axes to be `None` (in which case let's change the test). 

Also I think this is probably better 🤔 , I'm not 100% sure though. It feels somewhat arbitrary.. so I'd vote for going with it and we can always change these edge case behaviors if it makes sense later.

File: python/tests/test_vmap.py [hunk 2]
```diff
@@ -462,6 +462,21 @@ def scatter(a, idx, updates):
...
+        a = mx.random.uniform(shape=(2, 3, 4))
+        b = mx.random.uniform(shape=(4, 3))
+
+        def const_func(a, b):
+            return mx.array(2)
+
+        out = mx.vmap(const_func, in_axes=(0, None))(a, b)
+        self.assertTrue(mx.array_equal(mx.full((2,), 2), out))
+        out = mx.vmap(const_func, in_axes=(None, 0))(a, b)
+        self.assertTrue(mx.array_equal(mx.full((4,), 2), out))
```
Comment (awni) [inline python/tests/test_vmap.py:475]: What happens / should happen if you do?

```python
        out = mx.vmap(const_func, in_axes=(0, 0))(a, b)
```

File: mlx/transforms.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (barronalex) [reply mlx/transforms.cpp:654]: Yeah I agree I'm slightly ambivalent about it too but it is what people coming from Jax are expecting so I'm down to put it in.

**Review by awni — APPROVED** [review]
Thanks for the fix!!
